### PR TITLE
fix: pg span names

### PR DESCRIPTION
**Pull request by ethanresnick**
## Which problem is this PR solving?

The primary functional change in this PR is to address a number of issues with how the pg instrumentation was generating span names.

1. Before this PR, when the query being executed was a prepared statement with a `name`, the `name` was transformed first before it was put into the span's name. E.g., `client.query({ name: 'find user', text: 'SELECT * from user where id = ?', values: [1] })` would generate a span name of `pg.query:find`, because the string `'find user'` would be split on spaces and only the first word would be used. However, it's the full prepared statement name that constitutes the low cardinality identifier, so this commit now makes the span name: `pg.query:find user`.

2. Similarly, before this PR, the parsed out operation name wasn't normalized before using it in the span name. Accordingly, a `select * from x` would get a span name of `pg.query:select`, while a `SELECT * from x` query would generate `pg.query:SELECT`. Now, they'll both have the latter span name.

3. Previously, the generated span names did not include the DB name, although the examples in the otel spec recommend this. Now, the db name is included, as it's specified in the connection info.

There are also two commits prior to the commit that adjusts the span naming strategy. Those commits do not make any functional changes, but just refactor/DRY up the existing code, which allows the new span names to be constructed in only one place. They also make the logic a bit more defensive, so that various edge cases with invalid arguments (e.g., `client.query(null)` or `client.query(queryText, valuesThatArentAnArray)`) can no longer cause the instrumentation itself to fail with an uncaught exception (though errors raised synchronously or asynchronously from `client.query` are still propagated as before). 

## Short description of the changes

See each commit message and the overview above.

### CODE REVIEW

**Comment by ethanresnick** [comment]
@Rauno56 Re the CI fail: I'm happy to add more tests first before this can be merged, but, before I take the time to do that, please let me know if the PR looks like it's on the right track and is something you think will be accepted. Thanks!

File: plugins/node/opentelemetry-instrumentation-pg/src/utils.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-pg/src/utils.ts:80]: Now that `command` can contain spaces and the database name is also appended, it might create span names such as `pg.query:find users users` which I guess are not straight forward to read.

I read your awesome description above about the complexity of building the span name in this case. I am good with keeping it as it improves the current state, but wonder if you considered any alternatives which might be more readable to users or better align with the spec suggestion `<db.operation> <db.name>`

**Review by blumamir — COMMENTED** [review]
Big like on these refactors and the work for improving the span name and error handling.
Added a few minor comments but overall I support these changes.

Your PR makes some changes with no update to existing tests, so I guess the current test coverage is not great. I Would be happy to see some tests to assert the new logic :)

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:185]: Great catch! 

Technically, the code as-is wouldn't try to add the callback as an attribute (because we check that we're dealing with an array before setting the attribute, to handle cases like `client.query({ values: notAnArray })`), but it would pass it to the hook, and it's better not to do that. I'll push a fix!

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:375]: We can't, unfortunately, because `pg.QueryConfig` makes a bunch of other promises to TS that the code hasn't actually verified. E.g., it says that the `name` property will be a string or undefined, which isn't necessarily true. Making these minimal ad-hoc types might seem a bit redundant, but it makes it much less likely that we'll accidentally lie to TS about what we've checked (esp as these types evolve over time), and it doesn't create the need for extra boilerplate casts or anything, since TS will compare the types structurally.

Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:375]: Done!

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:266]: > Can we replace with e?.message?

I don't think that's technically safe, because`e` could be a primitive value, and using `?.` with a primitive on the LHS would throw. Of course, it's quite unlikely that the thrown value would be a primitive, but there's nothing stopping that from being the case, so I figured we might as well be defensive about it.

Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:266]: Ok, I made a utility function for this logic, to keep the main code path a little more readable, and so that I could reuse it in one other place where the code was assuming that the thrown value would be an `Error`.

File: plugins/node/opentelemetry-instrumentation-pg/src/utils.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/utils.ts:80]: Hmm, it's a good question about alternatives. I'm not sure I have any great ideas, except maybe to put the name in quotation marks? So we'd have span names like:

```
pq.query:SELECT api-server-db
pg.query:"find user by id" api-server-db
pg.query:"findUser" api-server-db
```

I'm not an Otel expert, though, so I'm not sure if quotation marks in the span name are legal or, even if legal, are likely to cause problems for implementations.

**Comment by ethanresnick** [comment]
@blumamir Ok, I think I've addressed all your comments and have added some tests as well!

File: plugins/node/opentelemetry-instrumentation-pg/src/utils.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-pg/src/utils.ts:358]: nit: this might be more appropriate in `internal-types.ts`

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (blumamir) [inline plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:271]: is it feasible to write a test for this case?
Looks like when [first argument to query is `undefined` or `null` will trigger it to throw](https://github.com/brianc/node-postgres/blob/27d612a2ac2df8737397019a5806f745f19b760e/packages/pg/lib/client.js#L507) and it will also assert that instrumentation does not throw itself before calling origianl

**Review by blumamir — APPROVED** [review]
Thanks, LGTM

added 2 optional comments

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts]: I think we tend to assume that thrown errors are not primitives in many other instrumentations, which simplifies the code, but I accepts your defensive approach :)

File: plugins/node/opentelemetry-instrumentation-pg/src/utils.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-pg/src/utils.ts]: Yeah I am not sure it's a good idea. the spec doesn't say a lot about it.
Let's keep it as is for now and change in the future if we get feedback

File: plugins/node/opentelemetry-instrumentation-pg/src/utils.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/utils.ts:358]: This type is pretty tightly-coupled to exactly what's being verified in the `isObjectWithTextString`, and we need to make sure the type and the predicate function stay in sync, so I think it makes sense to keep them together. Other code also probably shouldn't depend on this type directly; it's just here to help TS narrow. Once TS is updated, this type won't even need to be named or exported, as we'll be able to remove all the `as ObjectWithText` casts, because TS will correctly infer that automatically. I.e., it'll be possible to just have:

```ts
export function isObjectWithTextString(it: unknown): it is { text: string, [k: string]: unknown } {
  // ...
}
```

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (ethanresnick) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:271]: Yup, I added explicit tests now for `client.query(undefined)` and `client.query(null)`. I think that's what you were talking about? The code on main already had some tests that the error thrown is from the pg library (not the instrumentation) and that a span is generated in the error status. I suppose I could also asserting on the span's message, but I don't think we want to couple the tests to the exact wording of the pg package's current errors.

**Comment by ethanresnick** [comment]
@blumamir Thanks for the quick + detailed feedback! I added a couple more test cases per your comments. Is this good to merge now, or does it need a second reviewer?

File: plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts
Comment (blumamir) [reply plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:271]: Right, I didn't see it already had tests, sorry about that
I thought the code to capture `original` invocations exceptions is new, but now I see it was just moved from `handleInvalidQuery` function 💯

**Comment by blumamir** [comment]
> @blumamir Thanks for the quick + detailed feedback! I added a couple more test cases per your comments. Is this good to merge now, or does it need a second reviewer?

Thanks again for the contribution and great improvements 🥇 
This is good to merge as far as I am concerned. @Rauno56 is the component owner - are you interested in adding a review?

**Comment by Rauno56** [comment]
I don't want to be the blocker and don't have cycles to review it this week. Can merge if we have one pair of maintainer eyes seen and approved it.

**Comment by ethanresnick** [comment]
Thanks @Rauno56!

**Pull request merged**
